File: advancedscreenswitchersample/src/main/java/com/art/advancedscreenswitchersample/ui/MainActivity.java
package com.art.advancedscreenswitchersample.ui;

import android.os.Bundle;
import android.support.annotation.IdRes;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;

import com.art.alligator.NavigationContext;
import com.art.alligator.NavigationContextBinder;
import com.art.alligator.NavigationFactory;
import com.art.alligator.Navigator;
import com.art.alligator.TransitionAnimation;
import com.art.alligator.implementation.FragmentScreenSwitcher;
import com.art.advancedscreenswitchersample.R;
import com.art.advancedscreenswitchersample.SampleAnimationProvider;
import com.art.advancedscreenswitchersample.SampleApplication;
import com.art.advancedscreenswitchersample.screens.TabScreen;
import com.roughike.bottombar.BottomBar;
import com.roughike.bottombar.OnTabSelectListener;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Date: 21.01.2016
 * Time: 23:30
 *
 * @author Artur Artikov
 */
public class MainActivity extends AppCompatActivity implements OnTabSelectListener {
	private static final String ANDROID_SCREEN_NAME = "ANDROID";
	private static final String BUG_SCREEN_NAME = "BUG";
	private static final String DOG_SCREEN_NAME = "DOG";

	private Navigator mNavigator;
	private NavigationContextBinder mNavigationContextBinder;
	private TabsInfo mTabsInfo;
	private FragmentScreenSwitcher mScreenSwitcher;

	@BindView(R.id.activity_main_bottom_bar)
	BottomBar mBottomBar;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		mNavigator = SampleApplication.getNavigator();
		mNavigationContextBinder = SampleApplication.getNavigationContextBinder();
		setContentView(R.layout.activity_main);
		ButterKnife.bind(this);

		mBottomBar.setOnTabSelectListener(this, false);
		initTabsInfo();
		initScreenSwitcher();

		if (savedInstanceState == null) {
			mNavigator.switchTo(ANDROID_SCREEN_NAME);
		}
	}

	@Override
	protected void onResume() {
		super.onResume();
		bindNavigationContext();
	}

	@Override
	protected void onPause() {
		mNavigationContextBinder.unbind();
		super.onPause();
	}

	@Override
	public void onBackPressed() {
		mNavigator.goBack();
	}

	@Override
	public void onTabSelected(@IdRes int tabId) {
		mNavigator.switchTo(mTabsInfo.getScreenName(tabId));
	}

	private void initTabsInfo() {
		mTabsInfo = new TabsInfo();
		mTabsInfo.add(ANDROID_SCREEN_NAME, R.id.tab_android, new TabScreen(getString(R.string.tab_android)));
		mTabsInfo.add(BUG_SCREEN_NAME, R.id.tab_bug, new TabScreen(getString(R.string.tab_bug)));
		mTabsInfo.add(DOG_SCREEN_NAME, R.id.tab_dog, new TabScreen(getString(R.string.tab_dog)));
	}

	private void initScreenSwitcher() {
		mScreenSwitcher = new FragmentScreenSwitcher(getSupportFragmentManager(), R.id.activity_main_container) {
			@Override
			protected Fragment createFragment(String screenName) {
				NavigationFactory navigationFactory = SampleApplication.getNavigationFactory();
				return navigationFactory.createFragment(mTabsInfo.getScreen(screenName));
			}

			@Override
			protected TransitionAnimation getAnimation(String screenNameFrom, String screenNameTo) {
				int indexFrom = mTabsInfo.getTabIndex(screenNameFrom);
				int indexTo = mTabsInfo.getTabIndex(screenNameTo);
				if(indexTo > indexFrom) {
					return new TransitionAnimation(R.anim.slide_in_right, R.anim.slide_out_left);
				} else {
					return new TransitionAnimation(R.anim.slide_in_left, R.anim.slide_out_right);
				}
			}

			@Override
			protected void onScreenSwitched(String screenName) {
				bindNavigationContext();
				selectTab(mTabsInfo.getTabId(screenName));
			}
		};
	}

	private void bindNavigationContext() {
		Fragment fragment = mScreenSwitcher.getCurrentFragment();
		NavigationContext.Builder builder = new NavigationContext.Builder(this)
				.screenSwitcher(mScreenSwitcher)
				.animationProvider(new SampleAnimationProvider());

		if (fragment != null && fragment instanceof ContainerIdProvider) {
			int containerId = ((ContainerIdProvider) fragment).getContainerId();
			builder.fragmentManagerAndContainerId(fragment.getChildFragmentManager(), containerId);
		}
		mNavigationContextBinder.bind(builder.build());
	}

	private void selectTab(@IdRes int tabId) {
		mBottomBar.setOnTabSelectListener(null);    // workaround to don't call listener
		mBottomBar.selectTabWithId(tabId);
		mBottomBar.setOnTabSelectListener(this, false);
	}
}

File: advancedscreenswitchersample/src/main/java/me/aartikov/advancedscreenswitchersample/ui/MainActivity.kt
package me.aartikov.advancedscreenswitchersample.ui;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.view.MenuItem;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;

import com.google.android.material.bottomnavigation.BottomNavigationView;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import me.aartikov.advancedscreenswitchersample.R;
import me.aartikov.advancedscreenswitchersample.SampleApplication;
import me.aartikov.advancedscreenswitchersample.SampleScreenSwitcherAnimationProvider;
import me.aartikov.advancedscreenswitchersample.SampleTransitionAnimationProvider;
import me.aartikov.advancedscreenswitchersample.screens.MainScreen;
import me.aartikov.advancedscreenswitchersample.screens.TabScreen;
import me.aartikov.alligator.NavigationContext;
import me.aartikov.alligator.NavigationContextBinder;
import me.aartikov.alligator.Navigator;
import me.aartikov.alligator.Screen;
import me.aartikov.alligator.annotations.RegisterScreen;
import me.aartikov.alligator.listeners.ScreenSwitchingListener;
import me.aartikov.alligator.screenswitchers.FragmentScreenSwitcher;


@RegisterScreen(MainScreen.class)
public class MainActivity extends AppCompatActivity implements BottomNavigationView.OnNavigationItemSelectedListener, ScreenSwitchingListener {

	BottomNavigationView mBottomBar;

	private final Navigator mNavigator = SampleApplication.getNavigator();
	private final NavigationContextBinder mNavigationContextBinder = SampleApplication.getNavigationContextBinder();
	private FragmentScreenSwitcher mScreenSwitcher;

	@SuppressLint("UseSparseArrays")
	private Map<Integer, Screen> mTabScreenMap = new LinkedHashMap<>();

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_main);
		mBottomBar = findViewById(R.id.bottom_bar);

		initTabScreenMap();
		mBottomBar.setOnItemSelectedListener(this);
		mScreenSwitcher = new FragmentScreenSwitcher(SampleApplication.getNavigationFactory(), getSupportFragmentManager(),
				R.id.main_container, new SampleScreenSwitcherAnimationProvider(getTabScreens()));

		if (savedInstanceState == null) {
			mNavigator.switchTo(getTabScreen(R.id.tab_android));
		}
	}

	private void initTabScreenMap() {
		mTabScreenMap.put(R.id.tab_android, new TabScreen(getString(R.string.tab_android)));
		mTabScreenMap.put(R.id.tab_bug, new TabScreen(getString(R.string.tab_bug)));
		mTabScreenMap.put(R.id.tab_dog, new TabScreen(getString(R.string.tab_dog)));
	}

	@Override
	protected void onResumeFragments() {
		super.onResumeFragments();
		bindNavigationContext();
	}

	@Override
	protected void onPause() {
		mNavigationContextBinder.unbind(this);
		super.onPause();
	}

	private void bindNavigationContext() {
		NavigationContext.Builder builder = new NavigationContext.Builder(this, SampleApplication.getNavigationFactory())
				.screenSwitcher(mScreenSwitcher)
				.screenSwitchingListener(this)
				.transitionAnimationProvider(new SampleTransitionAnimationProvider());

		Fragment fragment = mScreenSwitcher.getCurrentFragment();
		if (fragment instanceof ContainerIdProvider) {
			builder.fragmentNavigation(fragment.getChildFragmentManager(), ((ContainerIdProvider) fragment).getContainerId());  // Use child fragment manager for nested navigation
		}

		mNavigationContextBinder.bind(builder.build());
	}

	@SuppressLint("MissingSuperCall")
	@Override
	public void onBackPressed() {
		mNavigator.goBack();
	}

	@Override
	public boolean onNavigationItemSelected(@NonNull MenuItem item) {
		Screen screen = getTabScreen(item.getItemId());
		mNavigator.switchTo(screen);
		return false;
	}

	@Override
	public void onScreenSwitched(@Nullable Screen screenFrom, @NonNull Screen screenTo) {
		int tabId = getTabId(screenTo);
		mBottomBar.getMenu().findItem(tabId).setChecked(true);
		bindNavigationContext();    // rebind NavigationContext because we need to set another container id and another child fragment manager.
	}

	private Screen getTabScreen(int tabId) {
		return mTabScreenMap.get(tabId);
	}

	private int getTabId(Screen tabScreen) {
		for (Map.Entry<Integer, Screen> entry : mTabScreenMap.entrySet()) {
			if (tabScreen.equals(entry.getValue())) {
				return entry.getKey();
			}
		}
		return -1;
	}

	private List<Screen> getTabScreens() {
		return new ArrayList<>(mTabScreenMap.values());
	}
}
